src/main/java/org/opensearch/security/user/UserService.java

```diff
@@ -0,0 +1,186 @@
+
+    final static String SERVICE_ACCOUNT_HASH_MESSAGE = "A password hash cannot be provided for service account. Failed to register service account: ";
+
+    final static String NO_ACCOUNT_NAME_MESSAGE = "No account name was specified in the request.";
+    private static CType getConfigName() {
+        return CType.INTERNALUSERS;
+    }
+
+    static final List<String> RESTRICTED_FROM_USERNAME = ImmutableList.of(
+            ":" // Not allowed in basic auth, see https://stackoverflow.com/a/33391003/533057
+    );
+
+    @Inject
+    public UserService(
+            ClusterService clusterService,
+            ConfigurationRepository configurationRepository,
+            Settings settings,
+            Client client
+    ) {
+        this.clusterService = clusterService;
+        this.configurationRepository = configurationRepository;
+        this.securityIndex = settings.get(ConfigConstants.SECURITY_CONFIG_INDEX_NAME,
+                ConfigConstants.OPENDISTRO_SECURITY_DEFAULT_CONFIG_INDEX);
+        this.client = client;
+    }
+
+    /**
+     * Load data for a given CType
+     * @param config CType whose data is to be loaded in-memory
+     * @return configuration loaded with given CType data
+     */
+    protected static final SecurityDynamicConfiguration<?> load(final CType config, boolean logComplianceEvent) {
+        SecurityDynamicConfiguration<?> loaded = configurationRepository.getConfigurationsFromIndex(Collections.singleton(config), logComplianceEvent).get(config).deepClone();
+        return DynamicConfigFactory.addStatics(loaded);
+    }
+
+    /**
+     * This function will handle the creation or update of a user account.
+     *
+     * @param accountDetailsAsString A string JSON of different account configurations.
+     * @return InternalUserConfiguration with the new/updated user
+     * @throws UserServiceException
+     * @throws IOException
+     */
+    public SecurityDynamicConfiguration<?> createOrUpdateAccount(String accountDetailsAsString) throws IOException {
+
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode accountDetails = mapper.readTree(accountDetailsAsString);
+        final ObjectNode contentAsNode = (ObjectNode) accountDetails;
+        SecurityJsonNode securityJsonNode = new SecurityJsonNode(contentAsNode);
+
+        final SecurityDynamicConfiguration<?> internalUsersConfiguration = load(getConfigName(), false);
+        String accountName = securityJsonNode.get("name").asString();
+
+        if (accountName == null || accountName.length() == 0) { // Fail if field is present but empty
+            throw new UserServiceException(NO_ACCOUNT_NAME_MESSAGE);
+        }
+
+        if (!securityJsonNode.get("attributes").get("owner").isNull() && !securityJsonNode.get("attributes").get("owner").equals(accountName)) { // If this is a service account
```

Comment:
Hi Craig, thank you for your good questions. Basically, the accountName is not an attribute. The accountName is the key to the user in the configuration. The owner is whoever 'owns' the account. For example, if you installed the HellowWolrld extension, it would make a service account for HelloWorld where you would be the owner even though the account name is HelloWorld. 

To your second point, the internalUsers are not the same as the User objects. The User objects that Ryan built the token generator around have a specific set of requirements for them that are more restrictive then the internalUsers. InternalUsers are mostly a data structure in practice and implementation. 

As a result of this, there are some differences with how the existing APIs were constructed. For instance, you can only specify four fields in the requests involving the internalUsers API one of which is the `attributes` field. If you were to try to specify individual fields the existing validator would need to be changed to accept them. The current implementation does allow an administrator to convert an account from a service account to normal account or vice versa. I don't know that this is necessarily a problem however.

Reply:
I reviewed the APIs and UI to ensure there is no avenue for a normal user to modify their own attributes. In general, it would be great to define a schema for what a service account entails vs. relying on keys to be present in a generic map of attributes.

A few of the concerns I have with this design are:

- An admin can at any time delete one of these `owner` or `isEnabled` keys and what would that affect be?
- If the user updates the value of the `accountName` to something that doesn't exist then there is no referential integrity
- I'm still not quite sure what `owner` means, you mentioned its a placeholder to provide the name of the person of initially installed an extension? 
- Attributes are read in from external IdPs or provided as a claim in a jwt, would this cause any problems?

Similar to what @RyanL1997 has done with support for extension REST handlers, would it make sense to merge a change like this into a feature branch while more of the feature is being worked on?

Reply:
Hi @cwperks,

Thank you for taking the time to look into this PR. I know you mentioned having some concerns about the implementation. Reviewing your concerns it seems they relate to the immutability of the service accounts. In general, it does not seem like there needs to be immutability of service accounts for several reasons. Fields such as `isEnabled` should be configurable by nature of what they represent. Another point is discussed in this comment [above](https://github.com/opensearch-project/security/pull/2594#discussion_r1160030423). Basically, it is not entirely clear the hard distinction between a normal internal user account and a service account. They are intended to be interchangeable for all intents and purposes and swapping between the two should not necessarily be forbidden. The owner denotes the controller of the account. If it is a user, it is their account. If instead, it is a service account, the service cannot "own" an account as they are just a program. The owner is then the person who installed that program. I don't see any reason why reading from a JWT would be an issue. 

To your point about merging against the feature/extensions branch, I am against this practice in general. I think that all changes which can be safely merged into the main branch should be. By merging into the feature branch, we add debt we will later have to address and make it more and more complicated to get an accurate state of the code base. 

If you would like to open up a PR swapping things to how you prefer, I would be happy to review the changes and make any switches that everyone prefers.

Reply:
@scrawfor99 If there's code that relies on attribute keys being present in this map then there needs to be special protections in place to ensure that the keys cannot just be deleted from the map as that would put the user/account into a corrupt state. I see that @peternied also had the same concerns: https://github.com/opensearch-project/security/pull/2594#discussion_r1160030423

Reply:
@cwperks, Peter's concern appears to have been about the meaning-fulllness of the terms not the inclusion of the terms themselves. I disagree about needing explicit safeguards for either `isEnabled` or `ownner`. If you look at the map now, there are numerous values that could be deleted or changed and cause a "corrupted state." I see no reason to treat these attributes any differently then something like `roles` which could be deleted to the result of having no access.

Reply:
@scrawfor99 What is preventing this scenario from happening/is this a concern with the design? From the design put forward for allowing extensions to write to reserved indices for the extension the extension would utilize a token associated with the (service) account designated to the extension. These indices come with special protections that even forbid an admin from deleting the index and rendering the extension into a corrupt state. However, with this design what is preventing this:

1. Admin errantly updates the attributes of the account associated with the extension and deletes the `owner` key from the attribute map
2. Since the `owner` key is no longer present the account associated with the extension is no longer one of these special accounts so its now forbidden from writing to the indices with system index protection.

Reply:
Hi @cwperks, there is nothing preventing an admin from deleting the `owner` or turning off `isEnabled`. This would make the service account no longer vend its associated token and it would no longer have access to the associated index. This is by design. I view this as an acceptable trade off for having service accounts be functionally identical to other internal user accounts. An admin can swap between the two, enable or disable both without hard deletion, and use the existing internaluser apis to access service accounts. For a P0 implementation these are the exact types of trade offs I am in favor of. We can always switch things down the road but these code changes introduce no immediate risks--service accounts are not used for anything currently and the api changes are not documented on the website--and are important for the extensions work.